verus/lcf/src/kernel.rs

```diff
@@ -0,0 +1,1108 @@
+use crate::string_hash_map::TmpStringHashMap;
+use vstd::assert_seqs_equal;
+use vstd::prelude::*;
+use vstd::seq::Seq;
+
+verus! {
+
+pub enum Const {
+    Atom(String),
+    Nat(u64),
+    Str(String),
+}
+
+impl PartialEq for Const {
+    //allows for Const types to be evaluated for equality and aid Verus verifier; not fully sufficient.
+    //Must call const_eq for Atoms and Strs
+    fn eq(&self, other: &Self) -> (res: bool)
+        ensures
+            res <==> self.deep_view() == other.deep_view(),
+    {
+        match (self, other) {
+            (Const::Atom(s), Const::Atom(t)) => s == t,
+            (Const::Nat(u), Const::Nat(v)) => u == v,
+            (Const::Str(s), Const::Str(t)) => s == t,
+            (Const::Atom(_), Const::Nat(_)) | (Const::Nat(_), Const::Atom(_)) => false,
+            (Const::Atom(_), Const::Str(_)) | (Const::Str(_), Const::Atom(_)) => false,
+            (Const::Str(_), Const::Nat(_)) | (Const::Nat(_), Const::Str(_)) => false,
+        }
+    }
+}
+
+impl Const {
+    // uses eq from impl Partial Eq to prove equality between items
```

Comment:
I would leave a comment here to say that this should eventually be replaced with the `impl PartialEq for Const`, once I can figure out what the issue is. you can even put `TODO(pratap)` in the comment